src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStackFrame.cs

```diff
@@ -89,6 +91,83 @@ private void WriteObjectOrArrayStart(ClassType classType, JsonEncodedText proper
             }
         }
 
+        public void WritePreservedObjectOrArrayStart(ClassType classType, Utf8JsonWriter writer, JsonSerializerOptions options, string referenceId)
+        {
+            if (JsonPropertyInfo?.EscapedName.HasValue == true)
+            {
+                WritePreservedObjectOrArrayStart(classType, JsonPropertyInfo.EscapedName!.Value, writer, referenceId);
+            }
+            else if (KeyName != null)
+            {
+                JsonEncodedText propertyName = JsonEncodedText.Encode(KeyName, options.Encoder);
+                WritePreservedObjectOrArrayStart(classType, propertyName, writer, referenceId);
+            }
+            else
+            {
+                if ((classType & (ClassType.Object | ClassType.Dictionary)) != 0)
+                {
+                    writer.WriteStartObject();
+                    writer.WriteString("$id", referenceId);
+                    StartObjectWritten = true;
+                }
+                else
+                {
+                    // Wrap array into an object with $id and $values metadata properties.
+                    Debug.Assert(classType == ClassType.Enumerable);
+                    writer.WriteStartObject();
+                    writer.WriteString("$id", referenceId);
+                    writer.WritePropertyName("$values");
+                    writer.WriteStartArray();
+                    WriteWrappingBraceOnEndPreservedArray = true;
+                }
+            }
+        }
+
+        private void WritePreservedObjectOrArrayStart(ClassType classType, JsonEncodedText propertyName, Utf8JsonWriter writer, string referenceId)
```

Comment:
nit: There is some code duplication happening in these methods. Any way to restructure it to avoid it, if possible?

Reply:
I didn't realize until you pointed out. Refactoring... thanks.